[quant][fx] Add support for GRU in fx graph mode quantization

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91976

Summary:
might be needed by a meta-internal use case

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_rnn

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][fx] Add support for GRU in fx graph mode quantization

Summary:
might be needed by a meta-internal use case

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_rnn

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented on torch/ao/nn/quantized/reference/modules/rnn.py:
```diff
@@ -488,3 +488,124 @@ def from_float(cls, mod, weight_qparams_dict):
+                if wn.startswith("weight"):
+                    weight_or_bias = get_quantized_weight(self, wn)
+                else:
+                    weight_or_bias = getattr(self, wn)
```
just curious - is it customary to not quantize bias at all?

jerryzh168 replied:
it's a design choice when we are designing eager mode quantization. in computation, bias is actually quantized to int32 with the scale of input and weight, but we only have this information in graph mode quantization, not in eager, so we have to keep bias fp32 and quantize it internally in the quantized linear operator

jcaip commented on torch/ao/nn/quantized/reference/modules/rnn.py:
```diff
@@ -488,3 +488,124 @@ def from_float(cls, mod, weight_qparams_dict):
+        else:
+            result = _VF.gru(input, batch_sizes, hx, self.get_flat_weights(), self.bias,
+                             self.num_layers, self.dropout, self.training, self.bidirectional)
+        output = result[0]
```
nit: does `output, hidden = result` work here?

jerryzh168 replied:
probably, this is just copied from forward function of GRU though, will add a comment in the beginning

jcaip reviewed (approved):
lgtm

jerryzh168 committed:
Update on "[quant][fx] Add support for GRU in fx graph mode quantization"


Summary:
might be needed by a meta-internal use case

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_rnn

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge -g

jerryzh168 commented:
@pytorchbot rebase

jerryzh168 committed:
Update on "[quant][fx] Add support for GRU in fx graph mode quantization"


Summary:
might be needed by a meta-internal use case

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_rnn

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for GRU in fx graph mode quantization"


Summary:
might be needed by a meta-internal use case

Test Plan:
python test/test_quantization.py TestQuantizeFxOps.test_rnn

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge -g
